tech_reports/LLMs/llms.md

```diff
@@ -73,13 +73,175 @@ Authors:
+
+```python
+input_memcfg = ttnn.create_sharded_memory_config(
+  (
+    batch_size, # The HEIGHT of a single shard
+    hidden_dim // core_grid.num_cores, # The WIDTH of a single shard
+  ),
+  core_grid, # Core grid to shard over (e.g. 8x2)
+  ttnn.ShardStrategy.WIDTH, # WIDTH sharding (as opposed to HEIGHT or BLOCK)
+  ttnn.ShardOrientation.ROW_MAJOR, # Shards are laid out in a row-major order over the core grid
+  use_height_and_width_as_shard_shape=True,
+)
+```
+Now that we know activations should be width-sharded, the only design decision to make is the `core_grid` on which to shard over. This is where you pay attention to 1) any constraints that an op might have on the input core grid, 2) how the input core grid affects the speed of the op, and 3) how the input core grid interplays with the output core grid.
+
+There are some cases where you don't need to create a specific sharded memory config. In these cases, you can instead pass one of the following:
+1. `ttnn.DRAM_MEMORY_CONFIG` when you just want DRAM interleaved.
+2. `ttnn.L1_MEMORY_CONFIG` when you want L1 interleaved.
+3. `ttnn.L1_WIDTH_SHARDED_MEMORY_CONFIG` when you want width-sharded and the op can infer the core grid and shard shape.
+
+As always, you should try running your `ttnn` op in a unit test with whichever settings you provide. You may find that the op produces incorrect outputs because it's missing some validation or different shard specs are used between input/output and the op itself (as TT-Metal matures, the sharding logic will get better at detecting these edge cases). You may also find that your memory config is not optimal and you can improve performance with a different configuration.
+
+Be careful when your memory config creates shards that require padding (i.e, the shard shape does not divide evenly into 32x32 tiles). Padded shards and padded ops are under active development and can be sources of bugs. When your memory config requires padding, you probably want to instead find a core grid which divides evenly into the tensor shape.
+
+#### Program Configs and Picking the Right Matmul
+Each `ttnn` operation has its own unique program config class. In general, program configs configure the op with hyperparameters that affects their functionality and performance. There are too many ops and program configs to cover in detail. We will focus on `ttnn.matmul` since it has multiple variants and it usually requires the most care.
+
+Picking a matmul variant is a key decision in optimizing a model. The choice depends on the shapes of the inputs and outputs and how the matmul fits into the rest of the model. You choose a variant by providing a specific `program_config` to `ttnn.matmul`. The following presents three matmul variants that are commonly used in LLMs.
+
+##### Matmul 2D
```

Comment:
Maybe we should take out Matmul 1D and 2D generic descriptions out of LLMs section, make their own tech report?

Reply:
I lean toward keeping them in this doc since they provide some background info for Ammar's MLP section, but we can always refactor or ask the tech writer to refactor generic sections out of this report.